README.md
# <a name="top"></a>xUI

* [概述](#概述)
* [安装](#安装)
* [使用](#使用)
* [xCollectionView](#xcollectionview)
   * [简单列表](#simplelist)
   * [多section列表](#sectionlist)
   * [向左靠齐布局](#alignleft)
   * [下拉刷新](#refresh)
   * [分页](#paging)
   * [拖拽排序](#drag)
* [xTableView](#xtableview)
   * [概述](#overview)
   * [左滑操作](#leftswipe)
* [xBanner](#xbanner)
   * [简单例子](#bannerexample)
   * [用作tab切换](#bannercomplicated)
   * [接口概览](#bannerinterface)
* [xPager](#xpager)
   * [用法举例](#xpagerexample)  
* [xWebView](#xwebview)
   * [基本用法](#xwebviewbasic)
   * [与JS交互](#xwebviewjs)
   * [定制ui](#xwebviewui)
* [xToast](#xtoast)
   * [用法举例](#xtoastexample)
* [xLoading](#xloading)
   * [用法举例](#xloadingexample)  
* [xAlert](#xalert)
   * [用法举例](#xalertexample) 
* [xCornerView](#xcornerview)
   * [用法举例](#xcornerexample) 
* [xGradientView](#xgradientview)
   * [用法举例](#xgradientexample) 
* [xBasePopView](#xbasepopview)
   * [单独使用](#xpopviewexample) 
   * [结合弹窗队列使用](#xpopviewqueue) 
* [xViewFactory](#xviewfactory)
   * [用法举例](#xviewfactoryexample) 

## 概述

业务无关的通用iOS UI组件库，支持ObjC & Swift

## 安装

通过pod引用，在podfile增加下面一行，通过tag指定版本
```
pod 'xUI',     :git => "https://github.com/jinsikui/xUI.git", :tag => 'v2.0.0-0'
```
 objc代码中引入：
```
#import <xUI/xUI.h>
```
 swift代码中引入：
```
在项目的plist文件中配置置 'Objective-C Bridging Header = xxx.h'，在xxx.h中添加一行：
#import <xUI/xUI.h>
```

## 使用

### xCollectionView 
[回顶](#top) 

```
封装了UICollectionView，使用丰富的属性和callback简化代码，支持下拉刷新，分页，拖拽排序等功能
```

#### <a name="simplelist"></a> - 简单列表

![Alt text](/Readme/pay-rank-list.jpeg)

```objc
// 创建组件
xCollectionView *cv = [[xCollectionView alloc] initWithCollectionViewCell]; // initWithCellClass:CHPositionCell.class

// 水平滚动（布局）
cv.scrollDirection = UICollectionViewScrollDirectionHorizontal; // 默认vertical
cv.clipsToBounds = false;

// 不允许用户滑动
cv.isScrollEnabled = false;

// 固定item size, 非固定size参考itemSizeCallback
cv.itemSize = CGSizeMake(28, 28);

// 水平滚动：列间距，垂直滚动：行间距
cv.lineSpace = 4;

// 创建cell callback
cv.buildCellCallback = ^(UICollectionViewCell *cell, xCollectionViewCellContext *context) {
   // cell.x_data is the element in cv.dataList
   LivePayRankItem *item = cell.x_data;
   //
   UIImageView *avatarView = [cell viewWithTag:PAYRANK_CELL_AVATAR_TAG];
   UIImageView *crownView = [cell viewWithTag:PAYRANK_CELL_CROWN_TAG];
   if (avatarView == nil) {
         avatarView = //...
         crownView = //...
         avatarView.tag = PAYRANK_CELL_AVATAR_TAG;
         crownView.tag = PAYRANK_CELL_CROWN_TAG;
         [cell addSubview:avatarView];
         [cell addSubview:crownView];
         // layout them ...
   }
   // refresh cell UI
   [avatarView lv_showAvatar:item.avatar width:27 placeholderImage:[LiveUser defaultAvatar]];
   crownView.image = [UIImage imageNamed:[NSString stringWithFormat:@"live-crown-%ld",(long)item.rank]];
};

// 设置数据
NSArray<LivePayRankItem*> *dataList = ...
cv.dataList = dataList;

// 每次dataList改变, 调reloadData()
[cv reloadData];
```

#### <a name="sectionlist"></a> - 多section列表

![Alt text](/Readme/section-list.jpeg)

```objc
// 创建组件
// 需要把所有section的cellClass放进一个数组传进去
xCollectionView *cv = [[xCollectionView alloc] initWithCellClasses:@[CHPositionCell.class]];
// 整体上下滚动
cv.scrollDirection = UICollectionViewScrollDirectionVertical;
__weak typeof(self) weakSelf = self;

// 构建cell UI
cv.buildCellCallback = ^(UICollectionViewCell * _Nonnull cell, xCollectionViewCellContext * _Nonnull context) {
   
   // 任意类型用户自定义section相关数据
   NSDictionary *dic = context.section.sectionData;
   CHPositionType type = [dic[@"type"] unsignedIntegerValue];
   
   // context.data: 每个cell对应的数据（ == cell.x_data）
   [((CHPositionCell *)cell) updateWithType:type model:context.data];
};

// 构建每个section的header UI
cv.buildHeaderCallback = ^UICollectionReusableView * _Nullable(UICollectionReusableView * _Nonnull view, NSInteger sectionIndex, xCollectionViewSection * _Nonnull section) {
   
   NSDictionary *dic = section.sectionData;
   CHPositionType type = [dic[@"type"] unsignedIntegerValue];
   
   if (type == CHPositionTypeHostin) {
         // 第一个section没有header
         return nil;
   } else {
         // 第二个section
         UILabel *headerLabel = [view viewWithTag:kCHMicPositionHeaderTag];
         if (!headerLabel) {
            headerLabel = [xViewFactory labelWithText:@"围观用户" font:[xFont semiboldPFWithSize:13] color:[xColor fromRGB:0x6F798E] alignment:NSTextAlignmentLeft];
            headerLabel.tag = kCHMicPositionHeaderTag;
            [view addSubview:headerLabel];
            [headerLabel mas_updateConstraints:^(MASConstraintMaker *make) {
               make.leading.mas_equalTo(view).offset(25);
               make.bottom.mas_equalTo(view);
            }];
         }
         return view;
   }
};
cv.itemSizeCallback = ^CGSize(xCollectionViewCellContext * _Nonnull context) {
   // cell size
   // ...
};
cv.headerSizeCallback = ^CGSize(xCollectionViewSection * _Nonnull sectionData, NSInteger sectionIndex) {
   // 每个section的header size
   // ...
};
cv.sectionInsetCallback = ^UIEdgeInsets(xCollectionViewSection * _Nonnull sectionData, NSInteger sectionIndex) {
   // 每个section内容的整体外边距
   // ...
};
cv.lineSpaceCallback = ^CGFloat(xCollectionViewSection * _Nonnull sectionData, NSInteger sectionIndex) {
   // cv.scrollDirection == .vertical时，上下两行cell之间的间距
   // ...
};
cv.interitemSpaceCallback = ^CGFloat(xCollectionViewSection * _Nonnull sectionData, NSInteger sectionIndex) {
   // cv.scrollDirection == .vertical时，左右两个cell之间的间距
   // ...
};

// 构造数据
NSMutableArray<xCollectionViewSection*> *array = [NSMutableArray array];

// 第一个section
xCollectionViewSection *hostinSection = [[xCollectionViewSection alloc] init];

// section中的列表数据
hostinSection.dataList = self.hostinUserList;

// 用户自定义的section数据
hostinSection.sectionData = @{@"type" : @(CHPositionTypeHostin)};

// 每个section可以设置不同的cellClass，xCollectionView.initWithCellClasses(:) 要把所有的cellClass放进数组传进去
hostinSection.cellClass = CHPositionCell.class;

[array addObject:hostinSection];


// 第二个section
xCollectionViewSection *audienceSection = [[xCollectionViewSection alloc] init];

// section中的列表数据
audienceSection.dataList = self.audienceList;

// 用户自定义的section数据
audienceSection.sectionData = @{@"type" : @(CHPositionTypeAudience)};

// 每个section可以设置不同的cellClass
audienceSection.cellClass = CHPositionCell.class;

[array addObject:audienceSection];

// 设置数据
cv.dataSectionList = array;
[cv reloadData]; // 显示多section列表
```

#### <a name="alignleft"></a> - 向左靠齐布局

UICollectionView的瀑布流布局，默认情况下，先计算每一行能放下多少个元素，然后把这些元素平均分布到这一行，如下图所示：

![Alt text](/Readme/search-history-wrong.jpeg)

而对于像搜索记录等需求，我们希望这些元素向左靠齐布局。xCollectionView 提供了支持：
```swift
// init中传入isLeftAlign = true 即可
self.historyView = xCollectionView.init(cellClasses: [TSSearchHistoryCell.self], isLeftAlign: true)
```
![Alt text](/Readme/search-history-right.jpeg)

#### <a name="refresh"></a> - 下拉刷新

![Alt text](/Readme/refresh.gif)

`xCollectionView`内部基于`MJRefresh`提供了下拉刷新功能，用法举例：

```objc
// 创建组件
- (void)createView {
   xCollectionView *cv = [[xCollectionView alloc] initWithCellClass:LiveQuitPopViewCell.class];
   self.cv = cv;
   // ......
   // 一旦设置，即开启下拉刷新，返回结果后会自动更新数据并reloadData
   cv.refreshCallback = ^FBLPromise<xCollectionViewPageResult *> * _Nullable{
      return [weakSelf refreshData];
   };
   // 下面这句：在设置了refreshCallback后再隐藏header，这样只能通过调代码refresh()来刷新，不能通过下拉刷新
   // cv.collectionView.mj_header.hidden = YES;

   // 调用refreshCallback，用返回的数据设置dataList然后reloadData()
   [self.cv refresh];
}

// 刷新获取数据的方法，返回FBLPromise
- (FBLPromise<xCollectionViewPageResult *> *)refreshData {
    __weak typeof(self) weakSelf = self;
    int pageSize = 10;
    return [LiveAPI.shared getRoomsWithstatus:LiveRoomStatusLiving page:1 pageSize:pageSize].then(^id (NSDictionary *ret) {
         weakSelf.page = 1; //记录当前是第几页
         // 构造返回数据
         xCollectionViewPageResult *result = [[xCollectionViewPageResult alloc] init];
         NSArray<NSDictionary*> *dataList = ret[@"items"];
         // 是否没有更多数据了
         result.isNoMoreData = dataList.count < pageSize;
         // 数据列表
         result.pageDataList = [dataList x_map:^(NSDictionary *dic) {
            return [[LiveRoomCellModel alloc] initWithDic:dic];
         }];
         return result;
    });
}

```

主要接口如下：

```objc

@interface xCollectionView : UIView

#pragma mark - 下拉刷新

/// 一旦设置，即开启下拉刷新，返回结果后会自动更新数据并reloadData
@property(nonatomic,copy) xCollectionViewRefreshCallback _Nullable refreshCallback;

/// 分页刷新组件类型，必须设置为MJRefreshHeader的子类否则设置无效，默认为+defaultRefreshHeaderClass
@property(nonatomic) Class refreshHeaderClass;

/// 默认为MJRefreshNormalHeader
@property(nonatomic,class) Class defaultRefreshHeaderClass;

// 可以通过代码触发refresh
-(FBLPromise<xCollectionViewPageResult*>*)refresh;

// ......
@end

```
其中最核心的`xCollectionViewRefreshCallback`定义如下

```objc

/// 同时用于下拉刷新和分页返回数据
@interface xCollectionViewPageResult : NSObject

/// 如果是下一页：返回数组中的数据会填充到最后一个section的dataList的尾部
/// 如果是刷新：返回的数组中数据会作为xCollectionView.dataList
@property(nonatomic,strong) NSArray *_Nullable pageDataList;

/// 如果是下一页：返回数组中的数据会作为section添加至dataSectionList尾部，不要忘记设置cellClass属性
/// 如果是刷新：返回数组中数据会作为xCollectionView.dataSectionList
@property(nonatomic,strong) NSArray<xCollectionViewSection*> *_Nullable pageSectionList;

/// footer是否变为'没有更多数据'状态
@property(nonatomic,assign) BOOL isNoMoreData;

/// 仅对refreshCallback有效，可用于当第一页就知道没有更多数据时隐藏分页控件（下拉刷新后可以改变）
@property(nonatomic,assign) BOOL shouldHideFooter;

/// 如果设置为true，下拉刷新或加载下一页时，控件不会自动设置/刷新数据，留给业务自己设置数据并调reloadData
@property(nonatomic,assign) BOOL ignoreRetData;

@end

/// 默认会自动调用reloadData
typedef FBLPromise<xCollectionViewPageResult*>*_Nullable(^xCollectionViewRefreshCallback)(void);

```

#### <a name="paging"></a> - 分页

![Alt text](/Readme/next-page.gif)

`xCollectionView`内部基于`MJRefresh`提供了分页功能，用法举例：

```objc
// 创建组件
- (void)createView {
   xCollectionView *cv = [[xCollectionView alloc] initWithCellClass:LiveQuitPopViewCell.class];
   self.cv = cv;
   // ......
   // 一旦设置，即开启分页
   cv.nextPageCallback = ^FBLPromise<xCollectionViewPageResult *> * _Nullable{
      return [weakSelf loadNextPage];
   };
}

// 获取下一页数据
- (FBLPromise<xCollectionViewPageResult *> *)loadNextPage {
    __weak typeof(self) weakSelf = self;
    int pageSize = 10;
    return [LiveAPI.shared getRoomsWithStatus:LiveRoomStatusLiving page:self.page+1 pageSize:pageSize].then(^id (NSDictionary *ret) {
         weakSelf.page += 1; //记录当前是第几页

         // 构造返回数据
         xCollectionViewPageResult *result = [[xCollectionViewPageResult alloc] init];
         NSArray<NSDictionary*> *dataList = ret[@"items"];
         // 是否没有更多数据了
         result.isNoMoreData = dataList.count < pageSize;
         // 一页的数据列表
         result.pageDataList = [dataList x_map:^(NSDictionary *dic) {
            return [[LiveRoomCellModel alloc] initWithDic:dic];
         }];
         return result;
    });
}

```

主要接口如下：

```objc

@interface xCollectionView : UIView

#pragma mark - 分页

/// 一旦设置，即开启分页
@property(nonatomic,copy) xCollectionViewNextPageCallback _Nullable nextPageCallback;

/// 分页刷新组件类型，必须设置为MJRefreshFooter的子类否则设置无效，默认为+defaultRefreshFooterClass
@property(nonatomic) Class refreshFooterClass;

/// 默认为MJRefreshAutoNormalFooter
@property(nonatomic,class) Class defaultRefreshFooterClass;

// ......

@end

```
其中最核心的`xCollectionViewRefreshCallback`定义如下

```objc

/// 同时用于下拉刷新和分页返回数据
@interface xCollectionViewPageResult : NSObject

/// 如果是下一页：返回数组中的数据会填充到最后一个section的dataList的尾部
/// 如果是刷新：返回的数组中数据会作为xCollectionView.dataList
@property(nonatomic,strong) NSArray *_Nullable pageDataList;

/// 如果是下一页：返回数组中的数据会作为section添加至dataSectionList尾部，不要忘记设置cellClass属性
/// 如果是刷新：返回数组中数据会作为xCollectionView.dataSectionList
@property(nonatomic,strong) NSArray<xCollectionViewSection*> *_Nullable pageSectionList;

/// footer是否变为'没有更多数据'状态
@property(nonatomic,assign) BOOL isNoMoreData;

/// 仅对refreshCallback有效，可用于当第一页就知道没有更多数据时隐藏分页控件（下拉刷新后可以改变）
@property(nonatomic,assign) BOOL shouldHideFooter;

/// 如果设置为true，下拉刷新或加载下一页时，控件不会自动设置/刷新数据，留给业务自己设置数据并调reloadData
@property(nonatomic,assign) BOOL ignoreRetData;

@end

/// 默认会自动调用reloadData
typedef FBLPromise<xCollectionViewPageResult*>*_Nullable(^xCollectionViewNextPageCallback)(void);

```

#### <a name="drag"></a> - 拖拽排序

![Alt text](/Readme/drag.gif)

用法举例（swift）:

```swift
func createView() {
   // 创建组件
   cv = xCollectionView.init(cellClass: TSPhotoCell.self)
   // ......
   // 一旦设置，即开启长按拖拽排序，移动过程中dataSectionList(dataList)中的顺序也会随时改变
   cv.moveCallback = { [weak self] moveData in
      self?.movePhoto(moveData)
   }
   // 最后一个cell是添加按钮，不参与拖拽排序
   cv.moveDisableItems = [self.btnModel]

   // 加载数据
   TSAPI.shared().getPhotosWithUserId(TSAuth.shared().userId!).__onQueue(DispatchQueue.main, then: {ret in
      var dataList: [TSUserPhotoModel] = []
      let arr = JSON.init(ret)["items"].array
      for item in arr {
         dataList.append(TSUserPhotoModel.init(json: item))
      }
      if arr.count < 9 {
         // 如果形象照小于9个，显示添加按钮
         dataList.append(self.btnModel)
      }
      // 设置数据
      self.cv.dataList = dataList
      self.cv.reloadData()
      return nil
   })
}

func movePhoto(_ data: xCollectionViewMoveData) {
   if data.action == xCollectionViewMoveActionEnd {
      let end = data.endIndexPath!
      let start = data.startIndexPath!
      // 没有移动
      if end.item == start.item {
         return
      }
      let model = data.sortedDataList[end.item] as! TSUserPhotoModel
      // 通知服务端排序改变
      TSAPI.shared().sortPhoto(model.id, position: end.item).__onQueue(DispatchQueue.main, then: { ret in
         // 调接口成功
         xToast.show("已保存", duration: 1)
         return nil
      }).__onQueue(DispatchQueue.main, catch: { error in
         // 调接口失败，回退
         self.cv.moveItem(from: end, to: start)
         xToast.show("接口失败", duration: 1)
      })
   }
}

```
主要接口如下：

```objc

@interface xCollectionView : UIView
// ......
#pragma mark - 拖拽排序

/// 一旦设置，即开启长按拖拽排序，移动过程中dataSectionList(dataList)中的顺序也会随时改变
@property(nonatomic,copy) xCollectionViewMoveCallback _Nullable moveCallback;

/// 拖拽排序中不参与拖拽的项目（必须是dataSectionList(dataList)中的对象）
@property(nonatomic,strong) NSArray *_Nullable moveDisableItems;

/// 暴露给外界的接口，sectionDataList(dataList)会跟随改变，可用来回退
-(void)moveItemFrom:(NSIndexPath*)from to:(NSIndexPath*)to;

// ......

@end

```
其中核心的`moveCallback`定义如下

```objc
/**
    整个拖拽过程以一个xCollectionViewMoveActionBegin开始，
    跟随多个xCollectionViewMoveActionSwap，
    以xCollectionViewMoveActionEnd结束
 */
typedef enum xCollectionViewMoveAction{
    xCollectionViewMoveActionBegin,
    xCollectionViewMoveActionSwap,
    xCollectionViewMoveActionEnd
} xCollectionViewMoveAction;

@interface xCollectionViewMoveData : NSObject

@property(nonatomic,assign) xCollectionViewMoveAction action;

/// 触发拖拽的cell的indexPath
@property(nonatomic,strong) NSIndexPath *startIndexPath;

/// 拖拽结束时的indexPath
@property(nonatomic,strong) NSIndexPath *_Nullable endIndexPath;

/// 当action==xCollectionViewMovementActionSwap时，swapFrom，swapTo都不为空，否则为空
@property(nonatomic,strong) NSIndexPath *_Nullable swapFrom;
@property(nonatomic,strong) NSIndexPath *_Nullable swapTo;

/// 按拖拽后新的顺序返回
@property(nonatomic,strong) NSArray *sortedDataList;
@property(nonatomic,strong) NSArray<xCollectionViewSection*> *sortedSectionDataList;

@end

typedef void(^xCollectionViewMoveCallback)(xCollectionViewMoveData*_Nonnull);

```


### xTableView
[回顶](#top)

#### <a name="overview"></a> - 概述

```
内部封装了UITableView，使用丰富的属性和callback简化代码，支持下拉刷新，分页，左划操作等功能
xTableView的使用方法和xCollectionView类似，主要的不同在于xTableView不支持水平列表，但支持每行的左划操作，比如左划删除行
```

#### <a name="leftswipe"></a> - 左划操作

![Alt text](/Readme/leftswipe.gif)

```objc

@interface xTableView : UIView

#pragma mark - 左滑操作

@property(nonatomic,copy) BOOL(^_Nullable canEditRowCallback)(xTableViewCellContext *context);

@property(nonatomic,copy) NSArray<UITableViewRowAction*>*_Nullable(^_Nullable editActionsForRowCallback)(xTableViewCellContext *context);

// ......

@end

```
用法举例（swift）:

```swift

func createUI() {
   tv = xTableView.init(cellClass: TSRoomRowCell.self)
   // ......
   tv.canEditRowCallback = {_ in
      return true
   }
   tv.editActionsForRowCallback = {_ in
      let del = UITableViewRowAction.init(style: .destructive, title: "取消收藏") { [weak self] (_, indexPath) in
            self?.delete(indexPath)
      }
      del.backgroundColor = TSColors.mainColor
      return [del]
   }
   // ......
}

func delete(_ indexPath: IndexPath) {
   let data = self.dataList[indexPath.item]
   TSAPI.shared().deleteData(data.roomId).__onQueue(DispatchQueue.main, then: {_ in
      // UI删除
      self.tv.deleteRow(at: indexPath)
      return nil
   }
}

```

### xBanner
[回顶](#top)

![Alt text](/Readme/banner1.gif)
![Alt text](/Readme/banner2.gif)
```
功能丰富的banner，支持水平和垂直布局，支持循环播放，自动播放，页码回调等功能
```

#### <a name="bannerexample"></a> - 简单例子
![Alt text](/Readme/banner2.gif)
```objc
// 创建控件
xBanner *banner = [[xBanner alloc] initWithCellClass:UICollectionViewCell.class itemSize:CGSizeMake(250, 350)];
[self.view addSubview:banner];
// banner布局 ......

// 开启自动播放
banner.isAutoScroll = true;

// 开启循环播放
banner.isCycleScroll = true;

// 自动播放每页停留时间
banner.autoScrollIntervalSeconds = 1.5f;

// 构建cell
banner.buildCellCallback = ^(UICollectionViewCell * _Nonnull cell, xBannerCellContext * _Nonnull context) {
   UIImageView *imgv = [cell.contentView viewWithTag:1024];
   if(!imgv){
      // 添加一个UIImageView
      imgv = [UIImageView new];
      imgv.clipsToBounds = true;
      imgv.contentMode = UIViewContentModeScaleAspectFill;
      imgv.tag = 1024;

      // 布局
      [cell.contentView addSubview:imgv];
      [imgv mas_makeConstraints:^(MASConstraintMaker *make) {
            make.edges.mas_equalTo(0);
      }];
   }
   [imgv sd_setImageWithURL:[NSURL URLWithString:(NSString*)cell.x_data]];
};
__weak typeof(self) weak = self;

// 页码回调
banner.indexChangeCallback = ^(UICollectionViewCell * _Nonnull cell, xBannerCellContext * _Nonnull context) {
   weak.pageLabel.text = [NSString stringWithFormat:@"%ld/3", context.dataIndex + 1];
};

// 设置数据
banner.dataList = @[@"https://tuboshu-static.oss-cn-beijing.aliyuncs.com/tuboshu-ios/stuffs/test-beauty.jpeg",
                  @"https://tuboshu-static.oss-cn-beijing.aliyuncs.com/tuboshu-ios/stuffs/test-beauty4.jpeg",
                  @"https://tuboshu-static.oss-cn-beijing.aliyuncs.com/tuboshu-ios/stuffs/test-beauty3.jpeg"];

// 显示数据
[banner reloadData];

// 页码label
_pageLabel = [UILabel new];
[self.view addSubview:_pageLabel];
// label布局 ......
```

#### <a name="bannercomplicated"></a> - 用作tab切换
```
xBanner常用于tab切换的控件，每个tab对应一个banner的cell，cell中是复杂的内容或者列表
此时应该禁止banner重用外层的cell：banner.reuseEnabled = false; 可以在页码回调中更新tab header UI
```

#### <a name="bannerinterface"></a> - 接口概览
```objc
@interface xBanner : UIView

/// 是否支持循环（仅当dataList.count > 1时才起作用）
@property(nonatomic) BOOL isCycleScroll;

/// 是否支持自动播放，仅当 isCycleScroll==YES 时才有意义
@property(nonatomic) BOOL isAutoScroll;

/// 自动播放时间间隔
@property(nonatomic) int autoScrollIntervalSeconds;

/// 是否允许用户手动滑（不影响自动播放的滑动）
@property(nonatomic) BOOL scrollEnabled;

/// 是否支持cell重用（有些情况下比如cell中是xCollectionView，不希望重用外面banner的cell）
@property(nonatomic) BOOL reuseEnabled;

/// 数据源
@property(nonatomic,nullable) NSArray *dataList;

/// 需要告知cell的尺寸，本控件只支持固定尺寸的cell，应当与banner的整体尺寸相同
@property(nonatomic) CGSize itemSize;

/// 水平 or 垂直 banner
@property(nonatomic) UICollectionViewScrollDirection scrollDirection;

/// banner的cell类型，需要是UICollectionViewCell或子类
@property(nonatomic) Class cellClass;

/// 构建cell回调
@property(nonatomic,nullable) xBannerCellCallback buildCellCallback;

/// 点击cell回调
@property(nonatomic,nullable) xBannerCellCallback selectCellCallback;

/// 稳定的停住后回调
@property(nonatomic,nullable) xBannerCellCallback stopToItemCallback;

/// 页码改变回调
@property(nonatomic,nullable) xBannerCellCallback indexChangeCallback;

/// 初始化
-(instancetype)initWithCellClass:(Class)cellClass itemSize:(CGSize)itemSize;

/// 通过代码触发banner滚动
-(void)scrollToDataIndex:(NSInteger)dataIndex animated:(BOOL)animated;

/// 显示banner数据
-(void)reloadData;

@end
```

### xPager
[回顶](#top)
```
可定制尺寸和颜色的圆形页码UI
```
#### <a name="xpagerexample"></a> - 用法举例
![Alt text](/Readme/pager.gif)
```objc

// 页码回调
banner.indexChangeCallback = ^(UICollectionViewCell * _Nonnull cell, xBannerCellContext * _Nonnull context) {
   // 设置当前页
   weak.pager.curPageIndex = context.dataIndex;
};

// 显示页码

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.0.0-0](https://github.com/jinsikui/xUI/compare/v1.0.0-1...v2.0.0-0) (2022-03-31)


### Features

* code complete ([07c1cfb](https://github.com/jinsikui/xUI/commit/07c1cfb48ad0488db67b858d3f3523e108b0b742))

## 1.0.0-1 (2022-03-30)


### Features

* setup project ([b0e9d44](https://github.com/jinsikui/xUI/commit/b0e9d443f7371a7db432fd5d33f6dd6418b97f6c))


xUI/MainViewController.m


#import "MainViewController.h"
#import "xUtil.h"
#import "xUI.h"
#import "Masonry.h"
#import "xBaseController.h"
#import "SDWebImage.h"
#import "TestBannerController.h"


@interface MainViewController ()
@property(nonatomic,strong) UIScrollView   *scroll;
@end

@implementation MainViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.title = @"xUI Test";
    self.view.backgroundColor = UIColor.whiteColor;
    _scroll = [[UIScrollView alloc] init];
    _scroll.alwaysBounceVertical = true;
    [self.view addSubview:_scroll];
    [_scroll mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.mas_equalTo(0);
    }];
    [self addBtn:@"9 patch" y:30 selector:@selector(action9Patch)];
    [self addBtn:@"alphaVideo-top" y:80 selector:@selector(actionAlphaVideoTop)];
    [self addBtn:@"alphaVideo-bottom" y:130 selector:@selector(actionAlphaVideoBottom)];
    [self addBtn:@"xBaseController" y:180 selector:@selector(baseControllerTest)];
    [self addBtn:@"loading" y:230 selector:@selector(loadingTest)];
    [self addBtn:@"toast" y:280 selector:@selector(toastTest)];
    [self addBtn:@"banner" y:330 selector:@selector(actionBanner)];
    [self addBtn:@"xAlert" y:380 selector:@selector(actionxAlert)];
    [self addBtn:@"xCornerView" y:430 selector:@selector(actionxCornerView)];
    [self addBtn:@"xGradientView" y:480 selector:@selector(actionGradientView)];
    //
    _scroll.contentSize = CGSizeMake(0, 580);
}

-(void)addBtn:(NSString*)text y:(CGFloat)y selector:(SEL)selector{
    UIButton *btn = [xViewFactory buttonWithTitle:text font:kFontRegularPF(12) titleColor:kColor(0) bgColor:xColor.clearColor borderColor:kColor(0) borderWidth:0.5];
    [btn addTarget:self action:selector forControlEvents:UIControlEventTouchUpInside];
    btn.frame = CGRectMake(0.5 * (xDevice.screenWidth - 150), y, 150, 35);
    [self.scroll addSubview:btn];
}

-(void)refreshConstraints{
    [self.view setNeedsUpdateConstraints];
    [self.view updateConstraintsIfNeeded];
}

#pragma mark - Actions

-(void)actionGradientView {
    xGradientView *gv = [[xGradientView alloc] init];
    gv.colors = @[
        (__bridge id)[xColor fromRGBA:0x0000FF alpha:0.76].CGColor,
        (__bridge id)[xColor fromRGBA:0x0000FF alpha:0].CGColor,
    ];
    // 从上到下渐变
    gv.startPoint = CGPointMake(0.5, 0);
    gv.endPoint = CGPointMake(0.5, 1);
    
    [self.view addSubview:gv];
    [gv mas_makeConstraints:^(MASConstraintMaker *make) {
        make.center.equalTo(self.view).offset(120);
        make.size.mas_equalTo(CGSizeMake(50, 25));
    }];
}

-(void)actionxCornerView {
    xCornerView *cv = [[xCornerView alloc] initWithCorners:UIRectCornerTopLeft|UIRectCornerBottomLeft radius:10];
    cv.backgroundColor = UIColor.systemBlueColor;
    [self.view addSubview:cv];
    [cv mas_makeConstraints:^(MASConstraintMaker *make) {
        make.center.equalTo(self.view).offset(120);
        make.size.mas_equalTo(CGSizeMake(50, 25));
    }];
}

-(void)actionxAlert {
    xAlert.presentFromControllerProvider = ^UIViewController * _Nullable{
        return UIApplication.sharedApplication.keyWindow.rootViewController;
    };
    [xAlert showSystemAlertWithTitle:@"提示" message:@"退出直播间会断开与主播的连麦" confirmText:@"再看看" cancelText:@"仍然退出" callback:^(UIAlertAction * _Nonnull action) {
        if([action.title isEqualToString:@"再看看"]){
            // 再看看 ...
        }
        else {
            // 仍然退出 ...
        }
    }];
}

-(void)actionBanner {
    [self.navigationController pushViewController:[TestBannerController new] animated:true];
}

-(void)actionAlphaVideoBottom{
    xAlphaVideoView *view = [xAlphaVideoView new];
    [self.view addSubview:view];
    [view playUrl:[NSBundle.mainBundle URLForResource:@"天秤" withExtension:@"mp4"] infoCallback:^(UIView * _Nonnull view, CGSize videoSize, double duration) {
        //这里不能直接引用self or self.view，会造成循环引用
        [view mas_makeConstraints:^(MASConstraintMaker *make) {
            make.width.mas_equalTo(kScreenWidth);
            make.height.mas_equalTo(kScreenWidth * videoSize.height / videoSize.width);
            make.centerX.mas_equalTo(0);
            make.bottom.mas_equalTo(0);
        }];
    } completion:^(UIView * _Nonnull view, NSError * _Nullable error) {
        NSLog(@"===== video complete: error(%@) =====", error);
        [view removeFromSuperview];
    }];
}

-(void)actionAlphaVideoTop{
    xAlphaVideoView *view = [xAlphaVideoView new];
    [self.view addSubview:view];
    [view playUrl:[NSBundle.mainBundle URLForResource:@"天秤" withExtension:@"mp4"] infoCallback:^(UIView * _Nonnull view, CGSize videoSize, double duration) {
        [view mas_makeConstraints:^(MASConstraintMaker *make) {
            make.width.mas_equalTo(kScreenWidth);
            make.height.mas_equalTo(kScreenWidth * videoSize.height / videoSize.width);
            make.centerX.mas_equalTo(0);
            make.top.mas_equalTo(0);
        }];
    } completion:^(UIView * _Nonnull view, NSError * _Nullable error) {
        NSLog(@"===== video complete: error(%@) =====", error);
        [view removeFromSuperview];
    }];
}

-(void)action9Patch{
    NSData *data = [xFile getDataFromFileOfPath:[xFile bundlePath:@"国王气泡.png"]];
    UIImage *resizableImg = [x9PatchParser resizableImageFromPngData:data scale:3];
    UIImageView *imgv = [[UIImageView alloc] initWithImage:resizableImg];
    [self.view addSubview:imgv];
    [imgv mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.mas_equalTo(100);
        make.top.mas_equalTo(200);
        make.width.mas_equalTo(200);
        make.height.mas_equalTo(150);
    }];
}

- (void)baseControllerTest {
    xBaseController * baseController = [xBaseController new];
    baseController.navigationBar.title = @"baseVC";
    baseController.navigationBar.leftText = @"返回";
    baseController.navigationBar.leftTextColor = [UIColor grayColor];
    baseController.navigationBar.leftBtnHandler = ^{
        [self.navigationController popViewControllerAnimated:true];
    };
    baseController.navigationBar.rightText = @"右侧返回";
    baseController.navigationBar.rightTextColor = [UIColor grayColor];
    baseController.navigationBar.rightBtnHandler = ^{
        [self.navigationController popViewControllerAnimated:true];
    };
    [baseController.navigationBar addBottomShadow];
    [self.navigationController pushViewController:baseController animated:true];
}

- (void)loadingTest {
    [xLoading showInWindow:@"加载中..."];
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [xLoading hide];
    });
}

- (void)toastTest {
    [xToast show:@"我是特别长的内容,我是特别长的内容,我是特别长的内容,我是特别长的内容,我是特别长的内容,我是特别长的内容,我是特别长的内容,我是特别长的内容,我是特别长的内容" duration:2];
}


@end

